`editor::SelectLargerSyntaxNode` crashes after exceeding excerpt boundaries

### Summary

<!-- Please insert a one line summary of the issue below -->

Crashes in multibuffer if `editor::SelectLargerSyntaxNode` exceeds the excerpt boundaries.

<!-- Include all steps necessary to reproduce from a clean Zed installation. Be verbose -->
Inexact steps to trigger:
 - Open reference of big Rust function in multibuffer, then trigger `editor::SelectLargerSyntaxNode` repeatedly.

Exact steps to trigger the problem:
1. Open Zed in the Zed repository
2. Run `git checkout e06666759a728503fed8b9f291dfea0b5f1eaf16` 
3. Go to `crates/editor/src/editor.rs:11113`
4. Ctrl-click on `go_to_diagnostic` to open multibuffer with references.
5. Trigger `editor: select larger syntax node` till it goes over excerpt boundaries
6. If you can't crash it, go to another reference of the same symbol and try again.
    - Rust-Analyzer order might not be deterministic.



Actual Behavior:

Crashes.

Expected Behavior:

Don't crash, not sure what the selection behavior should be tho, maybe it should no-op if the expansion would go over the excerpt boundary.


### Zed Version and System Specs

Zed: v0.174.6 (Zed) 
OS: Linux X11 arch unknown
Memory: 60.3 GiB
Architecture: x86_64
GPU: AMD Radeon RX 6750 XT (RADV NAVI22) || radv || Mesa 24.3.4-arch1.1

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2025-02-24T19:14:57.795560966-03:00 [ERROR] {
  "thread": "main",
  "payload": "offset is out of range",
  "location_data": {
    "file": "crates/editor/src/display_map/fold_map.rs",
    "line": 729
  },
  "backtrace": [
    "zed::reliability::init_panic_hook::{{closure}}::h8d31ed9815dfa227+118629683",
    "std::panicking::rust_panic_with_hook::hfe205f6954b2c97b+94028664",
    "std::panicking::begin_panic_handler::{{closure}}::h6cb44b3a50f28c44+94027827",
    "std::sys::backtrace::__rust_end_short_backtrace::hf1c1f2a92799bb0e+94020249",
    "rust_begin_unwind+94026996",
    "core::panicking::panic_fmt::h3d8fc78294164da7+7484707",
    "editor::display_map::fold_map::FoldSnapshot::intersects_fold::hde7185403da6dcf2+37576137",
    "editor::Editor::select_larger_syntax_node::h1c893ee37cce8f0f+39728093",
    "editor::element::register_action::{{closure}}::hbcc3eab83059cd9e+38304816",
    "gpui::window::Window::dispatch_action_on_node::hf9345c85bdf3bda5+54415872",
    "gpui::window::Window::dispatch_event::hb490813a6029f258+54489505",
    "gpui::window::Window::new::{{closure}}::h3e3829924c840740+54343988",
    "gpui::platform::linux::x11::window::X11WindowStatePtr::handle_input::hedfb0883c11104f4+54122374",
    "gpui::platform::linux::x11::client::X11Client::handle_event::ha87146a1f201863b+54062617",
    "gpui::platform::linux::x11::client::X11Client::process_x11_events::h082fd5fdca4d0686+54036876",
    "<core::cell::RefCell<calloop::sources::DispatcherInner<S,F>> as calloop::sources::EventDispatcher<Data>>::process_events::h219b6a5b1096a40a+50053563",
    "<gpui::platform::linux::x11::client::X11Client as gpui::platform::linux::platform::LinuxClient>::run::heb555db15c62e6df+54098838",
    "gpui::platform::linux::platform::<impl gpui::platform::Platform for P>::run::h82e07a1c5ed24f84+53713661",
    "gpui::app::Application::run::h574a3567f4b8ec86+116252102",
    "zed::main::h4f876862b654b2bf+118685207",
    "std::sys::backtrace::__rust_begin_short_backtrace::h08842ea0d3573fec+117668227",
    "std::rt::lang_start::{{closure}}::h0aad445293a7e5c5+119483321",
    "std::rt::lang_start_internal::h5e7c81cecd7f0954+93953397",
    "main+118700860",
    "__libc_start_main+30934851691852",
    "_start+11104110"
  ],
  "app_version": "0.174.6",
  "app_commit_sha": "d5fde6645afe6356ef31611a76aff69a3e450260",
  "release_channel": "stable",
  "target": "x86_64-unknown-linux-gnu",
  "os_name": "Linux X11",
  "os_version": "arch unknown",
  "architecture": "x86_64",
  "panicked_on": 1740435297795,
  "system_id": "5f06c84b-cb3d-4c89-88a6-31b82da45990",
  "installation_id": "a2a70d96-426e-446a-8f5e-1ba4ad0fbbbe",
  "session_id": "bc984a6d-d0b1-4320-963d-184b6fba5a8a"
}
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -132,7 +132,7 @@ pub use multi_buffer::{
 };
 use multi_buffer::{
     ExcerptInfo, ExpandExcerptDirection, MultiBufferDiffHunk, MultiBufferPoint, MultiBufferRow,
-    ToOffsetUtf16,
+    MultiOrSingleBufferOffsetRange, ToOffsetUtf16,
 };
 use project::{
     lsp_store::{CompletionDocumentation, FormatTrigger, LspFormatTarget, OpenLspBufferHandle},
@@ -10727,7 +10727,10 @@ impl Editor {
                 while let Some((node, containing_range)) = buffer.syntax_ancestor(new_range.clone())
                 {
                     new_node = Some(node);
-                    new_range = containing_range;
+                    new_range = match containing_range {
+                        MultiOrSingleBufferOffsetRange::Single(_) => break,
+                        MultiOrSingleBufferOffsetRange::Multi(range) => range,
+                    };
                     if !display_map.intersects_fold(new_range.start)
                         && !display_map.intersects_fold(new_range.end)
                     {

--- a/crates/editor/src/hover_popover.rs
+++ b/crates/editor/src/hover_popover.rs
@@ -15,7 +15,7 @@ use itertools::Itertools;
 use language::{DiagnosticEntry, Language, LanguageRegistry};
 use lsp::DiagnosticSeverity;
 use markdown::{Markdown, MarkdownStyle};
-use multi_buffer::ToOffset;
+use multi_buffer::{MultiOrSingleBufferOffsetRange, ToOffset};
 use project::{HoverBlock, HoverBlockKind, InlayHintLabelPart};
 use settings::Settings;
 use std::{borrow::Cow, cell::RefCell};
@@ -447,11 +447,13 @@ fn show_hover(
                     })
                     .or_else(|| {
                         let snapshot = &snapshot.buffer_snapshot;
-                        let offset_range = snapshot.syntax_ancestor(anchor..anchor)?.1;
-                        Some(
-                            snapshot.anchor_before(offset_range.start)
-                                ..snapshot.anchor_after(offset_range.end),
-                        )
+                        match snapshot.syntax_ancestor(anchor..anchor)?.1 {
+                            MultiOrSingleBufferOffsetRange::Multi(range) => Some(
+                                snapshot.anchor_before(range.start)
+                                    ..snapshot.anchor_after(range.end),
+                            ),
+                            MultiOrSingleBufferOffsetRange::Single(_) => None,
+                        }
                     })
                     .unwrap_or_else(|| anchor..anchor);
 

--- a/crates/multi_buffer/src/multi_buffer.rs
+++ b/crates/multi_buffer/src/multi_buffer.rs
@@ -79,6 +79,12 @@ pub struct MultiBuffer {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
+pub enum MultiOrSingleBufferOffsetRange {
+    Single(Range<usize>),
+    Multi(Range<usize>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub enum Event {
     ExcerptsAdded {
         buffer: Entity<Buffer>,
@@ -5682,13 +5688,19 @@ impl MultiBufferSnapshot {
     pub fn syntax_ancestor<T: ToOffset>(
         &self,
         range: Range<T>,
-    ) -> Option<(tree_sitter::Node, Range<usize>)> {
+    ) -> Option<(tree_sitter::Node, MultiOrSingleBufferOffsetRange)> {
         let range = range.start.to_offset(self)..range.end.to_offset(self);
         let mut excerpt = self.excerpt_containing(range.clone())?;
         let node = excerpt
             .buffer()
             .syntax_ancestor(excerpt.map_range_to_buffer(range))?;
-        Some((node, excerpt.map_range_from_buffer(node.byte_range())))
+        let node_range = node.byte_range();
+        let range = if excerpt.contains_buffer_range(node_range.clone()) {
+            MultiOrSingleBufferOffsetRange::Multi(excerpt.map_range_from_buffer(node_range))
+        } else {
+            MultiOrSingleBufferOffsetRange::Single(node_range)
+        };
+        Some((node, range))
     }
 
     pub fn outline(&self, theme: Option<&SyntaxTheme>) -> Option<Outline<Anchor>> {
@@ -6660,9 +6672,17 @@ impl<'a> MultiBufferExcerpt<'a> {
 
     /// Map a range within the [`Buffer`] to a range within the [`MultiBuffer`]
     pub fn map_range_from_buffer(&mut self, buffer_range: Range<usize>) -> Range<usize> {
+        if buffer_range.start < self.buffer_offset {
+            log::warn!("Attempting to map a range from a buffer offset that starts before the current buffer offset");
+            return buffer_range;
+        }
         let overshoot = buffer_range.start - self.buffer_offset;
         let excerpt_offset = ExcerptDimension(self.excerpt_offset.0 + overshoot);
         self.diff_transforms.seek(&excerpt_offset, Bias::Right, &());
+        if excerpt_offset.0 < self.diff_transforms.start().1 .0 {
+            log::warn!("Attempting to map a range from a buffer offset that starts before the current buffer offset");
+            return buffer_range;
+        }
         let overshoot = excerpt_offset.0 - self.diff_transforms.start().1 .0;
         let start = self.diff_transforms.start().0 .0 + overshoot;
 

